modules/OFConnectionManager2/module/src/cxn_instance.c

```diff
@@ -0,0 +1,2122 @@
+#define LOG_INFO(cxn, fmt, ...)                                 \
+    AIM_LOG_INFO("cxn %s: " fmt, cxn->desc, ##__VA_ARGS__)
+#define LOG_VERBOSE(cxn, fmt, ...)                              \
+    AIM_LOG_VERBOSE("cxn %s: " fmt, cxn->desc, ##__VA_ARGS__)
+#define LOG_TRACE(cxn, fmt, ...)                                \
+    AIM_LOG_TRACE("cxn %s: " fmt, cxn->desc, ##__VA_ARGS__)
+#define LOG_INTERNAL(cxn, fmt, ...)                             \
+    AIM_LOG_INTERNAL("cxn %s(%p): " fmt, cxn->desc, cxn, ##__VA_ARGS__)
+
+
+/* Forward declaractions */
+static void
+cxn_state_set(connection_t *cxn, cxn_state_t new_state);
+static indigo_error_t
+cxn_try_to_connect(connection_t *cxn);
+
+
+/* Maximum number of messages to send per write callback */
+#define MAX_WRITE_MSGS 32
+
+
+/**
+ * Connection control blocks, indexed by connection index
+ */
+static connection_t connection[MAX_CONNECTIONS];
+
+/****************************************************************
+ * Connection instance bookkeeping
+ ****************************************************************/
+
+#define ITERATE_OVER_ALL_CXNS(_idx, _cxn)                               \
+    for (_idx = 0, _cxn = &connection[0];                               \
+         _idx < MAX_CONNECTIONS;                                        \
+         ++_idx, _cxn = &connection[_idx])
+
+/* Includes local connection */
+#define FOREACH_ACTIVE_CXN(_idx, _cxn)                                  \
+    ITERATE_OVER_ALL_CXNS(_idx, _cxn)                                   \
+    if (CXN_ACTIVE(_cxn))
+
+/* Only remote connections */
+#define FOREACH_REMOTE_ACTIVE_CXN(_idx, _cxn)                           \
+    ITERATE_OVER_ALL_CXNS(_idx, _cxn)                                   \
+    if (CXN_ACTIVE(_cxn) && !CXN_LOCAL(_cxn))
+
+/* All remote main connections which completed hand-shake and with 
+ * requested role */
+#define FOREACH_MAIN_HS_COMPLETE_CXN_WITH_ROLE(_idx, _cxn, _cxn_role)   \
+    ITERATE_OVER_ALL_CXNS(_idx, _cxn)                                   \
+    if (CXN_ACTIVE(_cxn) && !CXN_LOCAL(_cxn) &&                         \
+        (_cxn->aux_id == 0) &&                                          \
+        (_cxn->controller->role == _cxn_role) &&                        \
+        ind_cxn_is_handshake_complete(_cxn))
+
+
+static connection_t *
+find_free_connection(void) {
+    int idx;
+    for (idx = 0; idx < MAX_CONNECTIONS; ++idx) {
+        if (!connection[idx].active) {
+            return &connection[idx];
+        }
+    }
+
+    return NULL;
+}
+
+static void
+init_single_instance(connection_t *cxn, indigo_cxn_id_t cxn_id)
+{
+    INDIGO_MEM_CLEAR(cxn, sizeof(connection_t));
+    cxn->cxn_id = cxn_id;
+    cxn->write_queue = bigring_create(WRITE_QUEUE_SIZE, NULL);
+}
+
+void
+ind_cxn_init_instances(void)
+{
+    int idx;
+
+    for (idx = 0; idx < MAX_CONNECTIONS; ++idx) {
+        /* implicitly zero generation id */
+        init_single_instance(&connection[idx], (indigo_cxn_id_t) idx);
```

Comment:
this code was fantastically bad.  i ran valgrind on the unit test and caught a lot of leaks.